docroot/modules/custom/mass_schema_metatag/mass_schema_web_page/src/DecisionTokenRenderer.php

```diff
@@ -0,0 +1,167 @@
+use Drupal\Core\File\FileUrlGeneratorInterface;
+use Drupal\node\NodeInterface;
+
+/**
+ * Generates field tokens for the Decision content type.
+ */
+final class DecisionTokenRenderer {
+
+  private EntityTypeManagerInterface $entityTypeManager;
+
+  private FileUrlGeneratorInterface $fileUrlGenerator;
+
+  /**
+   * Construct a new DecisionTokenRenderer.
+   */
+  public function __construct(EntityTypeManagerInterface $entityTypeManager, FileUrlGeneratorInterface $fileUrlGenerator) {
+    $this->entityTypeManager = $entityTypeManager;
+    $this->fileUrlGenerator = $fileUrlGenerator;
+  }
+
+  /**
+   * Return if a node can be processed by this class.
+   *
+   * @param \Drupal\node\NodeInterface $node
+   *   The node to check.
+   *
+   * @return bool
+   *   TRUE if this node can be processed into tokens, FALSE otherwise.
+   */
+  public function isRenderable(NodeInterface $node): bool {
+    return $node->bundle() === 'decision';
+  }
+
+  /**
+   * Return an array of generated token values for use in hook_tokens().
+   *
+   * @param array $tokens
+   *   The array of tokens as per hook_tokens().
+   * @param \Drupal\node\NodeInterface $node
+   *   The decision node to render tokens from.
+   *
+   * @throws \LogicException
+   *   Thrown if the node is not the right content type.
+   *
+   * @return array
+   *   The array of rendered tokens.
+   */
+  public function getTokens(array $tokens, NodeInterface $node): array {
+    $replacements = [];
+
+    if (!$this->isRenderable($node)) {
+      throw new \LogicException('isRenderable() must be called before calling getTokens()');
+    }
+
+    // Iterate through all the tokens.
+    foreach ($tokens as $name => $original) {
+      // Check for the [field_reference:entity:field_name] token.
+      if (strpos($name, ':') !== FALSE) {
+        [$name] = explode(':', $name);
+      }
+
+      // If the node entity doesn't have the field, continue.
+      if ($node->hasField($name)) {
+        $field = $node->get($name);
+        // Logic for handling entity reference fields.
+        if ($field instanceof EntityReferenceFieldItemListInterface && $this->fieldReferencesDocument($field)) {
+          $this->getEntityReferenceReplacements($field, $replacements, $original);
+        }
+      }
+
+    }
+
+    return $replacements;
+  }
+
+  /**
+   * Add file paths to the replacements array if the target is a document.
+   *
+   * @param \Drupal\Core\Field\EntityReferenceFieldItemListInterface $field
+   *   The field referencing documents.
+   * @param array $replacements
+   *   The current array of token replacements.
+   * @param string $original
+   *   The original token name.
+   */
+  private function getEntityReferenceReplacements(EntityReferenceFieldItemListInterface $field, array &$replacements, string $original): void {
+    $filepaths = $this->getFilepaths($field);
+    $replacements[$original] = json_encode($filepaths, JSON_THROW_ON_ERROR);
```

Comment:
While this should never fail since we control the data, as is it would _silently_ fail so I added this flag.